mesonbuild/coredata.py

```diff
@@ -553,36 +560,54 @@ def set_default_options(self, default_options, subproject, cmd_line_options):
 
         self.set_options(options, subproject)
 
-    def process_new_compilers(self, lang: str, comp, cross_comp, cmd_line_options):
+    def process_new_compilers(self, lang: str, comp, cross_comp, env):
         from . import compilers
+
         self.compilers[lang] = comp
-        # Native compiler always exist so always add its options.
-        new_options = comp.get_options()
         if cross_comp is not None:
             self.cross_compilers[lang] = cross_comp
-            new_options.update(cross_comp.get_options())
+
+        # Native compiler always exist so always add its options.
+        new_options_for_build = comp.get_options()
+        preproc_flags_for_build = comp.get_preproc_flags()
+        if cross_comp is not None:
+            new_options_for_host = cross_comp.get_options()
+            preproc_flags_for_host = cross_comp.get_preproc_flags()
+        else:
+            new_options_for_host = comp.get_options()
+            preproc_flags_for_host = comp.get_preproc_flags()
+
+        opts_machines_list = [
+            (new_options_for_build, preproc_flags_for_build, MachineChoice.BUILD),
+            (new_options_for_host, preproc_flags_for_host, MachineChoice.HOST),
+        ]
 
         optprefix = lang + '_'
-        for k, o in new_options.items():
-            if not k.startswith(optprefix):
-                raise MesonException('Internal error, %s has incorrect prefix.' % k)
-            if k in cmd_line_options:
-                o.set_value(cmd_line_options[k])
-            self.compiler_options.setdefault(k, o)
-
-        # Unlike compiler and linker flags, preprocessor flags are not in
-        # compiler_options because they are not visible to user.
-        preproc_flags = comp.get_preproc_flags()
-        preproc_flags = shlex.split(preproc_flags)
-        self.external_preprocess_args.setdefault(lang, preproc_flags)
+        for new_options, preproc_flags, for_machine in opts_machines_list:
+            for k, o in new_options.items():
+                if not k.startswith(optprefix):
+                    raise MesonException('Internal error, %s has incorrect prefix.' % k)
+                if k in env.properties[for_machine]:
+                    # Get from configuration files.
+                    o.set_value(env.properties[for_machine][k])
+                if env.machines.matches_build_machine(for_machine) \
+                   and k in env.cmd_line_options:
```

Comment:
the visual indent here is off, the second line of the if statement should be indented twice so it doesn't line up with the body:
```python
if (... and
        ...):
    thing()
```

Also, please consider using parens instead of \ for line breaks, their clearer and easy to extend later if need be.